File: engine/core/turn_order.py
"""Turn order helpers."""

from __future__ import annotations

from typing import Dict, List

from engine.core.state import UnitState


def build_turn_order(units: Dict[str, UnitState]) -> List[str]:
    # Higher initiative acts first; ties break by stable unit id.
    ordered = sorted(units.values(), key=lambda u: (-u.initiative, u.unit_id))
    return [u.unit_id for u in ordered]


def next_turn_index(current: int, turn_order_size: int) -> int:
    return (current + 1) % turn_order_size

File: src/engine/turnOrder.ts
/**
 * Turn order helpers.
 * Mirrors engine/core/turn_order.py
 */

import { UnitState } from "./state";

export function buildTurnOrder(units: Record<string, UnitState>): string[] {
  // Higher initiative acts first; ties break by stable unit id.
  const ordered = Object.values(units).sort((a, b) => {
    if (b.initiative !== a.initiative) return b.initiative - a.initiative;
    return a.unitId.localeCompare(b.unitId);
  });
  return ordered.map((u) => u.unitId);
}

export function nextTurnIndex(current: number, turnOrderSize: number): number {
  return (current + 1) % turnOrderSize;
}
